File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.ProtectionProxyExample/DocumentProxy.cs
namespace ProxyDesignPattern.ProtectionProxyExample;

/// <summary>
/// Protection Proxy - Controls access to the document based on user role.
/// 
/// Access Rules:
/// - Viewer: Can only Read
/// - Editor: Can Read and Write
/// - Admin:  Can Read, Write and Delete
/// </summary>
public class DocumentProxy : IDocument
{
    private readonly SensitiveDocument _realDocument;
    private readonly User _user;

    public DocumentProxy(SensitiveDocument document, User user)
    {
        _realDocument = document;
        _user = user;
        Console.WriteLine($"[Proxy] Access granted to user '{_user.Name}' with role '{_user.Role}'");
    }

    public void Read()
    {
        // All roles can read
        Console.WriteLine($"[Proxy] User '{_user.Name}' is reading...");
        _realDocument.Read();
    }

    public void Write(string content)
    {
        // Only Editor and Admin can write
        if (_user.Role == Role.Viewer)
        {
            Console.WriteLine($"[Proxy] ACCESS DENIED: '{_user.Name}' (Viewer) cannot write!");
            return;
        }

        Console.WriteLine($"[Proxy] User '{_user.Name}' is writing...");
        _realDocument.Write(content);
    }

    public void Delete()
    {
        // Only Admin can delete
        if (_user.Role != Role.Admin)
        {
            Console.WriteLine($"[Proxy] ACCESS DENIED: '{_user.Name}' ({_user.Role}) cannot delete!");
            return;
        }

        Console.WriteLine($"[Proxy] User '{_user.Name}' is deleting...");
        _realDocument.Delete();
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.ProtectionProxyExample/src/DocumentProxy.ts
import { IDocument } from "./IDocument";
import { Role, User } from "./User";
import { SensitiveDocument } from "./SensitiveDocument";

/**
 * Protection Proxy - Controls access to the document based on user role.
 *
 * Access Rules:
 * - Viewer: Can only Read
 * - Editor: Can Read and Write
 * - Admin:  Can Read, Write and Delete
 */
export class DocumentProxy implements IDocument {
  constructor(
    private readonly _realDocument: SensitiveDocument,
    private readonly _user: User
  ) {
    console.log(`[Proxy] Access granted to user '${_user.name}' with role '${_user.role}'`);
  }

  read(): void {
    // All roles can read
    console.log(`[Proxy] User '${this._user.name}' is reading...`);
    this._realDocument.read();
  }

  write(content: string): void {
    // Only Editor and Admin can write
    if (this._user.role === Role.Viewer) {
      console.log(`[Proxy] ACCESS DENIED: '${this._user.name}' (Viewer) cannot write!`);
      return;
    }

    console.log(`[Proxy] User '${this._user.name}' is writing...`);
    this._realDocument.write(content);
  }

  delete(): void {
    // Only Admin can delete
    if (this._user.role !== Role.Admin) {
      console.log(`[Proxy] ACCESS DENIED: '${this._user.name}' (${this._user.role}) cannot delete!`);
      return;
    }

    console.log(`[Proxy] User '${this._user.name}' is deleting...`);
    this._realDocument.delete();
  }
}
